pixeltable/catalog/catalog.py

```diff
@@ -1579,44 +1594,72 @@ def get_tbl_version(
         # we need a transaction here, if we're not already in one; if this starts a new transaction,
         # the returned TableVersion instance will not be validated
         with self.begin_xact(for_write=False) as conn:
-            tv = self._tbl_versions.get((tbl_id, effective_version))
+            tv = self._tbl_versions.get(key)
             if tv is None:
-                tv = self._load_tbl_version(tbl_id, effective_version, check_pending_ops=check_pending_ops)
+                tv = self._load_tbl_version(key, check_pending_ops=check_pending_ops)
             elif not tv.is_validated:
                 # only live instances are invalidated
-                assert effective_version is None
-                # we validate live instances by comparing our cached TableMd.current_version/view_sn to what's stored
+                assert key.effective_version is None
                 # _logger.debug(f'validating metadata for table {tbl_id}:{tv.version} ({id(tv):x})')
-                q = sql.select(schema.Table.md).where(schema.Table.id == tbl_id)
-                row = conn.execute(q).one_or_none()
-                if row is None:
-                    raise excs.Error(self._dropped_tbl_error_msg(tbl_id))
-                current_version, view_sn = row.md['current_version'], row.md['view_sn']
+
+                reload = False
+
+                if tv.anchor_tbl_id is None:
+                    # live non-replica table; compare our cached TableMd.current_version/view_sn to what's stored
+                    q = sql.select(schema.Table.md).where(schema.Table.id == key.tbl_id)
+                    row = conn.execute(q).one_or_none()
+                    if row is None:
+                        raise excs.Error(self._dropped_tbl_error_msg(key.tbl_id))
+                    current_version, view_sn = row.md['current_version'], row.md['view_sn']
+                    if current_version != tv.version or view_sn != tv.tbl_md.view_sn:
+                        _logger.debug(
+                            f'reloading metadata for live table {key.tbl_id} '
+                            f'(cached/current version: {tv.version}/{current_version}, '
+                            f'cached/current view_sn: {tv.tbl_md.view_sn}/{view_sn})'
+                        )
+                        reload = True
+
+                else:
+                    # live replica table; use the anchored version
+                    anchor_tbl_md = self.head_version_md(tv.anchor_tbl_id)
```

Comment:
anchor_tbl_md is confusing because this isn't a TableMd instance.

Reply:
fixed